letsencrypt/error_handler.py

```diff
@@ -0,0 +1,98 @@
+"""Registers functions to be called if an exception or signal occurs."""
+import logging
+import os
+import signal
+import traceback
+
+
+logger = logging.getLogger(__name__)
+
+
+# _SIGNALS stores the signals that will be handled by the ErrorHandler. These
+# signals were chosen as their default handler terminates the process and could
+# potentially occur from inside Python. Signals such as SIGILL were not
+# included as they could be a sign of something devious and we should terminate
+# immediately.
+_SIGNALS = ([signal.SIGTERM] if os.name == "nt" else
+            [signal.SIGTERM, signal.SIGHUP, signal.SIGQUIT,
+             signal.SIGXCPU, signal.SIGXFSZ, signal.SIGPWR])
+
+
+class ErrorHandler(object):
+    """Registers functions to be called if an exception or signal occurs.
+
+    This class allows you to register functions that will be called when
+    an exception or signal is encountered. The class works best as a
+    context manager. For example:
+
+    with ErrorHandler(cleanup_func):
+        do_something()
+
+    If an exception is raised out of do_something, cleanup_func will be
+    called. The exception is not caught by the ErrorHandler. Similarly,
+    if a signal is encountered, cleanup_func is called followed by the
+    previously registered signal handler.
+
+    Every registered function is attempted to be run to completion
+    exactly once. If a registered function raises an exception, it is
+    logged and the next function is called. If a (different) handled
+    signal occurs while calling a registered function, it is attempted
+    to be called again by the next signal handler.
```

Comment:
So you might end up running registered function twice, which contradicts the first statement. This happens after exception, when registered function is processed and a signal arrives before it completes. The signal will call the same registered function once again (because `pop` happens later), but the first run of the function was never "stopped" as I understand it. Maybe you actually want to `pop` before calling?